contracts/zero-ex/contracts/src/features/libs/LibLimitOrder.sol

```diff
@@ -0,0 +1,208 @@
+    }
+
+    // The type hash for limit orders, which is:
+    // keccak256(abi.encodePacked(
+    //     "LimitOrder(",
+    //       "address makerToken,",
+    //       "address takerToken,",
+    //       "uint128 makerAmount,",
+    //       "uint128 takerAmount,",
+    //       "uint128 takerTokenFeeAmount,",
+    //       "address maker,",
+    //       "address taker,",
+    //       "address sender,",
+    //       "address feeRecipient,",
+    //       "bytes32 pool,",
+    //       "uint64 expiry,",
+    //       "uint256 salt"
+    //     ")"
+    // ))
+    uint256 private constant _LIMIT_ORDER_TYPEHASH =
+        0xce918627cb55462ddbb85e73de69a8b322f2bc88f4507c52fcad6d4c33c29d49;
+
+    // The type hash for RFQ orders, which is:
+    // keccak256(abi.encodePacked(
+    //     "RfqOrder(",
+    //       "address makerToken,",
+    //       "address takerToken,",
+    //       "uint128 makerAmount,",
+    //       "uint128 takerAmount,",
+    //       "address maker,",
+    //       "address txOrigin,",
+    //       "bytes32 pool,",
+    //       "uint64 expiry,",
+    //       "uint256 salt"
+    //     ")"
+    // ))
+    uint256 private constant _RFQ_ORDER_TYPEHASH =
+        0xc6b3034376598bc7f28b05e81db7ed88486dcdb6b4a6c7300353fffc5f31f382;
+
+    /// @dev Get the struct hash of a limit order.
+    /// @param order The limit order.
+    /// @return structHash The struct hash of the order.
+    function getLimitOrderStructHash(LimitOrder memory order)
+        internal
+        pure
+        returns (bytes32 structHash)
+    {
+        // The struct hash is:
+        // keccak256(abi.encode(
+        //   TYPE_HASH,
+        //   order.makerToken,
+        //   order.takerToken,
+        //   order.makerAmount,
+        //   order.takerAmount,
+        //   order.takerTokenFeeAmount,
+        //   order.maker,
+        //   order.taker,
+        //   order.sender,
+        //   order.feeRecipient,
+        //   order.pool,
+        //   order.expiry,
+        //   order.salt,
+        // ))
+        assembly {
+            let mem := mload(0x40)
+            mstore(mem, _LIMIT_ORDER_TYPEHASH)
+            // order.makerToken;
+            mstore(add(mem, 0x20), mload(order))
+            // order.takerToken;
+            mstore(add(mem, 0x40), mload(add(order, 0x20)))
+            // order.makerAmount;
+            mstore(add(mem, 0x60), mload(add(order, 0x40)))
+            // order.takerAmount;
+            mstore(add(mem, 0x80), mload(add(order, 0x60)))
+            // order.takerTokenFeeAmount;
+            mstore(add(mem, 0xA0), mload(add(order, 0x80)))
+            // order.maker;
+            mstore(add(mem, 0xC0), mload(add(order, 0xA0)))
+            // order.taker;
+            mstore(add(mem, 0xE0), mload(add(order, 0xC0)))
+            // order.sender;
+            mstore(add(mem, 0x100), mload(add(order, 0xE0)))
+            // order.feeRecipient;
+            mstore(add(mem, 0x120), mload(add(order, 0x100)))
```

Comment:
I think we may have to `and` addresses to mask off the upper bytes. Solidity can leave junk in those bits, or it might even be possible for someone to maliciously get them there on purpose via a `LimitOrder` struct passed via call data. I'm honestly not sure when we're guaranteed those bits are zeroed and when we're not.

Reply:
Oh, and same thing below in `getRfqOrderStructHash()`.

Reply:
Hmm, I wonder if we would ever actually take advantage of the dirty bits to embed extra data that could be part of the hash. I guess there's also an argument for having dirty bits that don't get hashed for injecting metadata post-signing.